# Table of Contents

## Importing Libraries

## Importing Dataframes

## Data Consistency Checks 
### Products 
#### - Missing Values for Products 
#### - Duplicates for Products 
#### - Export changes made to Products
### Orders
#### - Missing Values for Orders
#### - Duplicate Values for Orders
#### - Export changes made to Orders

----

# Importing Libraries

In [1]:
import pandas as pd
import numpy as ny
import os

# Importing Dataframes

In [2]:
path= r'C:\Users\princess\Documents\09-2023 Instacart Basket Analysis'

In [21]:
# Import wrangledorders.csv
df_ords=pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'),index_col=False)

In [4]:
# Import products.csv
df_prods=pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'),index_col=False)

In [5]:
# Import departments.csv
df_deps=pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'departments.csv'),index_col=False)

# Data Consistency Checks

## Products

### Missing Values for Products

In [6]:
#find missing values of df_prods dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [7]:
#create subset without the null values
df_nan= df_prods[df_prods['product_name'].isnull()==True]

In [8]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [9]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [10]:
df_prods.shape

(49693, 5)

In [11]:
# Clean products data
df_prods_clean= df_prods[df_prods['product_name'].isnull()==False]

In [12]:
df_prods_clean.shape

(49677, 5)

After removing the 16 missing values, we now have 49,677 rows.

### Duplicates for Products

In [19]:
# New dataframe with duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [14]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [15]:
df_prods_clean.shape

(49677, 5)

In [18]:
# Removing the duplicates from clean data and making new dataframe
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [17]:
df_prods_clean_no_dups.shape

(49672, 5)

After removing the duplicates, the rows went down to 49,672.

### Export changes made to Products

In [25]:
# Export changes 
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))

## Orders

In [22]:
#looking to see if anthing looks off about the df_ords dataframe

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

The days since last order column has a different count than all the rest of the columns. That might need to be investigated.
Additionally, the 50th percentile of 7 days fora max of 30 days doesn't seem right.

In [33]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  
0                  8                    NaN  
1                  7                   15.0  
2                 12                   21.0  
3                  7                   29.0  
4                 15                   28.0

In [31]:
#dropping eval_set 
df_ords=df_ords.drop(columns=['eval_set'])

In [24]:
# Check for mixed data types

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Nothing comes up when this code is run, meaning there are no mixed data types.

## Missing Values for Orders

In [34]:
# findings missing values 
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

Days since last order has a large amount of missing values, with 206,209. Since there are so many, creating a new data frame
would be best.

In [35]:
# new data frame subset for missing values in days since last order
df_ords_nan=df_ords[df_ords['days_since_last_order'].isnull()==True]

In [36]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

The total user_id's with missing values is equal to the number of missing data rows, which is 206,209.The order number for all of the customers with missing data, meaning that they would not have any prior orders to begin with. Therefore, I will leave it alone as that is still valuable data.

## Duplicate Values for Orders

In [37]:
# finding duplicate values for orders
df_ords_dups= df_ords[df_ords.duplicated()]

In [38]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

This shows that there are no duplicates in the ords dataframe. No changes need to be made.

### Export changes made to Orders

In [39]:
df_ords.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_checked.csv'))